libc/src/stdio/baremetal/printf.cpp

```diff
@@ -0,0 +1,51 @@
+//===-- Implementation of printf for baremetal ------------------*- C++ -*-===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+
+#include "src/stdio/printf.h"
+#include "src/__support/arg_list.h"
+#include "src/stdio/printf_core/core_structs.h"
+#include "src/stdio/printf_core/printf_main.h"
+#include "src/stdio/printf_core/writer.h"
+
+#include <stdarg.h>
+
+// This is intended to be provided by the vendor. Name TBD
+extern "C" size_t __llvm_libc_raw_write(const char *s, size_t size);
```

Comment:
How is this different from the `write_to_stderr` hook we already use? I guess the size is different. For the GPU I use `FILE` as a simple opaque handle basically, don't know if that's relevant here.

Reply:
I'd say `raw_write` and `write_to_stderr` are (in my mind) different in the same way `stdout` and `stderr` are. The intent for `stdout` is for normal output, whereas the intent for `stderr` is error messages. That being said, I should probably move this to `OSUtil` so that it's in the same place as `write_to_stderr`.

As for using `FILE` as an opaque handle, that probably would work, but the request from @petrhosek was for a printf that only writes to the provided function.

Reply:
`write_to_stderr` is a private C++ API, for vendor integration we need an internal C API. So far we only have `__llvm_libc_log_write` defined in baremetal `OSUtil`, and this change introduces `__llvm_libc_raw_write`. It's a question whether this API should be specific only to baremetal target or should be made more generic, but I'm fine starting with just baremetal and generalizing it as needed. It'd be also helpful to define all internal API in a single header that's extensively documented. There are also open questions around naming, versioning, prior art, etc. This is a broader topic that we should probably discuss on Discourse, but I'm fine landing this change in the meantime and refining the API later.

Regarding opaque `FILE`, for baremetal I think we'll need both. Most baremetal targets have some support for writing out characters, typically over serial. This would typically be used to implement `printf` (or similar API for logging). Many baremetal targets also support [semihosting](https://github.com/ARM-software/abi-aa/blob/5c0a393f50e083ccca9f32ca94995211141fe858/semihosting/semihosting.rst) which allows file system access and I'd like to implement `FILE` for baremetal through semihosting. We could also use semihosting to implement `printf` (using [`SYS_WRITE`](https://github.com/ARM-software/abi-aa/blob/5c0a393f50e083ccca9f32ca94995211141fe858/semihosting/semihosting.rst#sys_write-0x05) syscall) but that has a downside of requiring a debugger and so I don't think it should be the default implementation.

Reply:
having thought about it some more, I think the best solution might be to add a `write_to_stdout` that can be used by not just `printf` but also `puts` and `putchar` in future. I'll update this PR with what I currently have and then try making a version with this new design so that we can compare.

Reply:
I messed around with the design a bit and realized it's out of scope for this patch. I've filed an issue to track this further (https://github.com/llvm/llvm-project/issues/94685), but for now I'm going to land this patch as-is.